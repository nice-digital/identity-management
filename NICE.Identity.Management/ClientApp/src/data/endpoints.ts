const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
