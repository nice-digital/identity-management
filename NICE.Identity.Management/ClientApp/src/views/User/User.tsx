import React, { Component } from "react";

import { RouteComponentProps, Link, Redirect } from "react-router-dom";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";

import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";
import { UserType } from "../../models/types";
import { Endpoints } from "../../data/endpoints";
import { UnlockUser } from "../../components/UnlockUser/UnlockUser";
import { ResendVerification } from "../../components/ResendVerification/ResendVerification";
import { UserStatus } from "../../components/UserStatus/UserStatus";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { ToFormattedDateString } from "../../helpers/dateHelpers";

import styles from "./User.module.scss";
import { Button } from "@nice-digital/nds-button";

type TParams = { id: string };

type UserProps = Record<string, unknown> & RouteComponentProps<TParams>;

type UserState = {
	user: UserType;
	error?: Error;
	redirect: boolean;
	isLoading: boolean;
};

export class User extends Component<UserProps, UserState> {
	constructor(props: UserProps) {
		super(props);

		this.state = {
			user: {} as UserType,
			redirect: false,
			isLoading: true,
		};
	}

	handleError = (error: Error): void => {
		this.setState({ error });
	};

	updateData = (updatedData: UserType): void => {
		if (!Object.keys(updatedData).length) {
			this.setState({ redirect: true });
		} else {
			this.setState({ user: updatedData });
		}
	};

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const user = await fetchData(Endpoints.user(this.props.match.params.id));

		if (isDataError(user)) {
			this.setState({ error: user });
		}

		this.setState({ user, isLoading: false });
		document.title = `NICE Accounts - ${user.firstName} ${user.lastName}`;
	}

	render(): JSX.Element {
		const { user, error, redirect, isLoading } = this.state;
		const showResendVerificationButton = !user.hasVerifiedEmailAddress;

		if (redirect) {
			return <Redirect to="/users" />;
		}

		let lastBreadcrumb = `${user.firstName} ${user.lastName}`;

		if (isLoading) {
			lastBreadcrumb = "Loading user details";
		}
		if (error) {
			lastBreadcrumb = "Error";
		}

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb
						data-qa-sel="breadcrumb-user-link"
						to="/users"
						elementType={Link}
					>
						Users
					</Breadcrumb>
					<Breadcrumb>{lastBreadcrumb}</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader
							preheading="User profile for"
							heading={
								isLoading
									? "User details"
									: `${user.firstName} ${user.lastName}`
							}
							cta={
								<Button
									data-qa-sel="add-user-role-button"
									variant="cta"
									to={`/users/${this.props.match.params.id}/services`}
									elementType={Link}
									disabled={isLoading}
								>
									{isLoading ? "Loading..." : "Add role"}
								</Button>
							}
						/>
						<Grid>
							<GridItem cols={12} md={9} aria-busy={isLoading}>
								{isLoading ? (
									<p>Loading...</p>
								) : (
									<>
										<div className={`${styles.summaryList} pv--c`}>
											<span className={styles.summaryListLabel}>
												Account information
											</span>
											<div className={styles.summaryListDetail}>
												<div>
													<UserStatus user={user} />
												</div>
												{showResendVerificationButton && (
													<ResendVerification
														nameIdentifier={user.nameIdentifier}
														onError={this.handleError}
													/>
												)}

												<UnlockUser
													id={user.userId}
													isLocked={user.isLockedOut}
													onToggleLock={this.updateData}
													onError={this.handleError}
												/>
											</div>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Email address
											</span>
											<span className={styles.summaryListDetail}>
												{user.emailAddress}
											</span>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Initial registration date
											</span>
											<span className={styles.summaryListDetail}>
												{ToFormattedDateString(user.initialRegistrationDate)}
											</span>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Last logged in date
											</span>
											<span className={styles.summaryListDetail}>
												{ToFormattedDateString(user.lastLoggedInDate)}
											</span>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Name identifier (internal id)
											</span>
											<span className={styles.summaryListDetail}>
												{user.nameIdentifier}
											</span>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Is migrated from NICE Accounts
											</span>
											<span className={styles.summaryListDetail}>
												{user.isMigrated ? "Yes" : "No"}
											</span>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Is in Authentication Provider (Auth0)
											</span>
											<span className={styles.summaryListDetail}>
												{user.isInAuthenticationProvider ? "Yes" : "No"}
											</span>
										</div>

										<hr className="mv--b" />

										<h2 className="h3">Permanently delete this account</h2>
										<p>
											The account will no longer be available, and all data in
											the account will be permanently deleted.
										</p>
										<Link
											data-qa-sel="delete-user-link"
											to={`/users/${user.userId}/delete`}
										>
											Delete user
										</Link>
									</>
								)}
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
