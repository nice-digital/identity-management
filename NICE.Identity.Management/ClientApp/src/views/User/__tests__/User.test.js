import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";

import { User } from "../User";
import { Endpoints } from "../../../data/endpoints";
import singleUser from "./singleUser.json";
import { nextTick } from "../../../utils/nextTick";

describe("User", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = shallow(<User match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetch during componentDidMount", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = mount(<MemoryRouter><User match={match} /></MemoryRouter>);
		const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
		wrapper.instance().componentDidMount();
		wrapper.update();
		expect(spy).toHaveBeenCalled();
		expect(fetch.mock.calls.length).toEqual(1);
		expect(fetch.mock.calls[0][0]).toEqual(Endpoints.user(match.params.id));
		spy.mockClear();
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = shallow(<User match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetch returns 401 error", async () => {
		console.error = jest.fn();		
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<User match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();		
	});

	it("should show error message when fetch returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(
			<MemoryRouter>
				<User match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});
});
