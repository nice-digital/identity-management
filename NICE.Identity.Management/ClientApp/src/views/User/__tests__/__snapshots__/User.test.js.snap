// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      data-qa-sel="breadcrumb-user-link"
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb>
      Kristin Patrick
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    cta={
      <React.Fragment>
        <Button
          data-qa-sel="add-user-role-button"
          disabled={false}
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/users/1/services"
          variant="cta"
        >
          Add role
        </Button>
        <Button
          data-qa-sel="edit-profile-button"
          disabled={false}
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/users/1/edit"
        >
          Edit profile
        </Button>
      </React.Fragment>
    }
    heading="Kristin Patrick"
    preheading="User profile for"
  />
  <Grid
    elementType="div"
    equalHeight={false}
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      aria-busy={false}
      cols={12}
      md={9}
    >
      <div
        className="summaryList pv--c"
      >
        <span
          className="summaryListLabel"
        >
          Account information
        </span>
        <div
          className="summaryListDetail"
        >
          <div>
            <UserStatus
              user={
                Object {
                  "acceptedTerms": true,
                  "allowContactMe": true,
                  "emailAddress": "kristin.patrick@example.com",
                  "firstName": "Kristin",
                  "hasVerifiedEmailAddress": true,
                  "initialRegistrationDate": "2019-01-23T12:29:54.911Z",
                  "isLockedOut": false,
                  "isMigrated": false,
                  "isStaffMember": false,
                  "lastLoggedInDate": "2019-01-23T12:29:54.911Z",
                  "lastName": "Patrick",
                  "nameIdentifier": "auth0|abc12345-2f95-43e7-b419-18f2784f1111",
                  "userId": 1,
                }
              }
            />
          </div>
          <UnlockUser
            id={1}
            isLocked={false}
            onError={[Function]}
            onToggleLock={[Function]}
          />
        </div>
      </div>
      <div
        className="summaryList pv--c mb--d"
      >
        <span
          className="summaryListLabel"
        >
          Email address
        </span>
        <span
          className="summaryListDetail"
        >
          kristin.patrick@example.com
        </span>
      </div>
      <div
        className="summaryList pv--c mb--d"
      >
        <span
          className="summaryListLabel"
        >
          Initial registration date
        </span>
        <span
          className="summaryListDetail"
        >
          23 January 2019 12:29
        </span>
      </div>
      <div
        className="summaryList pv--c mb--d"
      >
        <span
          className="summaryListLabel"
        >
          Last logged in date
        </span>
        <span
          className="summaryListDetail"
        >
          23 January 2019 12:29
        </span>
      </div>
      <div
        className="summaryList pv--c mb--d"
      >
        <span
          className="summaryListLabel"
        >
          Name identifier (internal id)
        </span>
        <span
          className="summaryListDetail"
        >
          auth0|abc12345-2f95-43e7-b419-18f2784f1111
        </span>
      </div>
      <div
        className="summaryList pv--c mb--d"
      >
        <span
          className="summaryListLabel"
        >
          Is migrated from NICE Accounts
        </span>
        <span
          className="summaryListDetail"
        >
          No
        </span>
      </div>
      <div
        className="summaryList pv--c mb--d"
      >
        <span
          className="summaryListLabel"
        >
          Is in Authentication Provider (Auth0)
        </span>
        <span
          className="summaryListDetail"
        >
          No
        </span>
      </div>
      <hr
        className="mv--b"
      />
      <h2
        className="h3"
      >
        Permanently delete this account
      </h2>
      <p>
        The account will no longer be available, and all data in the account will be permanently deleted.
      </p>
      <Link
        data-qa-sel="delete-user-link"
        to="/users/1/delete"
      >
        Delete user
      </Link>
    </GridItem>
  </Grid>
</Fragment>
`;

exports[`User should show error message when fetch returns 401 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            data-qa-sel="breadcrumb-user-link"
            href="/users"
            onClick={[Function]}
          >
            Users
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Error
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/users\\",
        \\"name\\": \\"Users\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Error\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Error
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;

exports[`User should show error message when fetch returns 500 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            data-qa-sel="breadcrumb-user-link"
            href="/users"
            onClick={[Function]}
          >
            Users
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Error
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/users\\",
        \\"name\\": \\"Users\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Error\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Error
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;
